gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -248,24 +250,28 @@ private WriteObjectResponse doResumableUpload() throws IOException {
 
       boolean objectFinalized = false;
       while (!objectFinalized) {
-        WriteObjectRequest insertRequest;
-        if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
-          insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
+        WriteObjectRequest insertRequest = null;
+        if (requestChunkMap.size() > 0 && requestChunkMap.lastKey() >= writeOffset) {
+          insertRequest = getCachedRequest(requestChunkMap, writeOffset);
           writeOffset += insertRequest.getChecksummedData().getContent().size();
         } else {
-          ByteString data =
-              ByteString.readFrom(
-                  ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE), MAX_BYTES_PER_MESSAGE);
-          dataChunkMap.put(writeOffset, data);
-          if (dataChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
-            dataChunkMap.remove(dataChunkMap.firstKey());
+          if (requestChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
+            freeUpCommittedRequests(requestChunkMap, writeOffset);
+          } else {
```

Comment:
Considering that this path is most accessed, can we ensure that it is the first conditional ? avoiding two failed `if` checks for most calls ?

Reply:
I tried. The logic is not very readable. The logic for first 2 conditions are quite infrequent but to the point. this else is more of a fallback. I'll leave this as is in favor of code readability.

Reply:
We have atleast 95% of requests following this codepath, we could extract the conditional into a one-line method for readability. The cost might outway in this scenario, we could add additional comments and methods wherever possible

Refer to https://github.com/GoogleCloudDataproc/hadoop-connectors/pull/606 for how much conditionals can add to cpu cost.

Reply:
#606 is not relatable here. In #606 moreServerContent is an IO cost that we could have done without spending.

Here, the `if` condition will be called in each loop. It does not have any IO cost, it is just checking the occupancy of the buffer. What we save over here by re-ordering the code is `jmp` assembly instruction. I don't think we are doing a `assembly code` level optimization here.

To further illustrate if the code is like
```
while (cond) {
  if (cond2) {
    ... do x 5 % of the time ...
  } else {
   ... do y 95% of the time ...
  }
}
```

and if we try re-writing it as 
```

while (cond) {
  if (!cond2) {
   ... do y 95% of the time ...
  } else {
    ... do x 5 % of the time ...
  }
}
```

We do not save any CPU overhead. `cond2` will always be evaluated. What we will save here is below.

CPU will pipeline the assembly instructions to be executed even before evaluating `cond2`. If the condition evaluates to false, it flushes the pipeline and reloads it with the else block. 

This kind of optimization helps tremendously while writing kernel code or tight loops with huge iteration (like memcpy). It does not help us here.